### PR TITLE
Cleanup General Convolution

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
- Fixes a bug in our blurring/sharpening convolution algorithms where edges were not being blurred properly in images with transparency.
- Fixes a bug where convolution area lower bounds was not calculated properly.
- Minor performance updates and internal API cleanup.

### CODE REVIEW

File: src/ImageSharp/Processing/Processors/Convolution/EdgeDetectorCompassProcessor.cs [hunk 1]
```diff
@@ -104,7 +99,7 @@ protected override void OnFrameApply(ImageFrame<TPixel> source, Rectangle source
             // we need a clean copy for each pass to start from
             using (ImageFrame<TPixel> cleanCopy = source.Clone())
             {
-                new ConvolutionProcessor<TPixel>(kernels[0]).Apply(source, sourceRectangle, configuration);
+                new ConvolutionProcessor<TPixel>(kernels[0], true).Apply(source, sourceRectangle, configuration);
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/Processors/Convolution/EdgeDetectorCompassProcessor.cs]: If `preserveAlpha` is always true, I wonder if we need it actually.

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Convolution/EdgeDetectorCompassProcessor.cs]: Only for this algorithm. Other convolvers might not want it. 2 pass blur for example does not.

File: src/ImageSharp/Common/Helpers/DenseMatrixUtils.cs [hunk 2]
```diff
@@ -15,108 +16,126 @@ namespace SixLabors.ImageSharp
...
+                    vectorY += matrixY[y, x] * currentColor;
                 }
             }
 
-            ref Vector4 target = ref targetRow[column];
-            vector.W = target.W;
+            var vector = Vector4.SquareRoot((vectorX * vectorX) + (vectorY * vectorY));
+            ref Vector4 target = ref Unsafe.Add(ref targetRowRef, column);
+
+            if (preserveAlpha)
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/DenseMatrixUtils.cs]: What I’m finding with these algorithms is that there always seems to be something that comes up that requires a per pixel operation. I’m hoping the new simd APIs will offer ways to improve our options.

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/DenseMatrixUtils.cs]: I've created separate methods now to remove the branching. Much cleaner.

**Comment by JimBobSquarePants** [comment]
Merging this. I want to move on.

**Pull request merged**